experimental/virtio/src/vsock/packet.rs

```diff
@@ -0,0 +1,299 @@
+            anyhow::bail!("data length does not match the paacket's configured payload length");
+        }
+
+        self.buffer[HEADER_SIZE..(HEADER_SIZE + len as usize)].copy_from_slice(data);
+        Ok(())
+    }
+
+    fn read_u16(&self, offset: usize) -> u16 {
+        let mut data = [0; size_of::<u16>()];
+        data.copy_from_slice(&self.buffer[offset..(offset + size_of::<u16>())]);
+        u16::from_le_bytes(data)
+    }
+
+    fn read_u32(&self, offset: usize) -> u32 {
+        let mut data = [0; size_of::<u32>()];
+        data.copy_from_slice(&self.buffer[offset..(offset + size_of::<u32>())]);
+        u32::from_le_bytes(data)
+    }
+
+    fn read_u64(&self, offset: usize) -> u64 {
+        let mut data = [0; size_of::<u64>()];
+        data.copy_from_slice(&self.buffer[offset..(offset + size_of::<u64>())]);
+        u64::from_le_bytes(data)
+    }
+
+    fn write_u16(&mut self, offset: usize, value: u16) {
+        let dest = &mut self.buffer[offset..(offset + size_of::<u16>())];
+        dest.copy_from_slice(&value.to_le_bytes()[..]);
+    }
+
+    fn write_u32(&mut self, offset: usize, value: u32) {
+        let dest = &mut self.buffer[offset..(offset + size_of::<u32>())];
+        dest.copy_from_slice(&value.to_le_bytes()[..]);
+    }
+
+    fn write_u64(&mut self, offset: usize, value: u64) {
+        let dest = &mut self.buffer[offset..(offset + size_of::<u64>())];
+        dest.copy_from_slice(&value.to_le_bytes()[..]);
+    }
+}
+
+/// Vsock Ops.
+#[derive(Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u16)]
+pub enum VSockOp {
+    Invalid = 0,
+    /// Connection request.
+    Request = 1,
+    /// Connections accepted response.
+    Response = 2,
+    /// Connection reset, either in reponse to a shutdown request or invalid packets received.
+    Rst = 3,
+    /// Connection shutdown request.
+    Shutdown = 4,
+    /// Represents a data packet.
+    ///
+    /// Used to send payload. Only data packets with this op can contain a payload.
+    Rw = 5,
+    /// Give update on credit to support flow control, either in response to a credit request or at
+    /// any time it might be useful to inform the peer of the state of the stream buffer.
+    CreditUpdate = 6,
+    /// Request for update on credit to calculate stream buffer availability.
+    CreditRequest = 7,
+}
+
+bitflags! {
+    /// Flags about a socket connection.
+    ///
+    /// Used in a connection shutdown request.
+    /// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-4070005>.
+    pub struct VSockFlags: u32 {
+        /// Indicates that no more payload data will be received.
+        const NO_RECEIVE = 1;
+        /// Indicates that no more payload data will be senn.
+        const NO_SEND = 2;
+    }
+}
+
+/// Socket Type.
+#[derive(Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u16)]
+pub enum VSockType {
+    /// Only stream sockets are currently supported.
```

Comment:
Might be nice to clarify that Only stream sockets are currently supported in the spec. (as opposed to in our implementation). :)